docs/details/vision.dox

```diff
@@ -133,20 +133,70 @@ equal to the number of features contained in the query array.
-distances of the query array and another containing the indices of these
-distances in the training array. The resulting 1-dimensional arrays have length
-equal to the number of features contained in the query array.
+\brief Determine the nearest neighbouring points to a given set of points
+
+In this function, a "point" is simply a geometric point's coordinates in an
+n-dimensional space. The coordinates are described along dim0 (much like a
+traditional column vector). The \p train and \p query arrays then, are each
+expected to be a list of such coordinates, enumerated along dim1 (the usage of
+dim0 and dim1 here can also be switched, as explained later).
+
+This function then calculates which points in the \p train array are nearest to
+each point in the \p query array, based on the distance metric specified by \p
+dist_type: \ref AF_SAD (sum of absolute differences), \ref AF_SSD (sum of
+squared differences), or \ref AF_SHD (hamming distance).
+
+The number of nearest neighboring points returned is specified by \p n_dist,
+and the results are described in two output arrays:
+- \p idx:  contains the index of each result that corresponds to the point in
+           \p train
+- \p dist: contains the distance from the query point to the result's
+           corresponding point in \p train
+
+In both the output arrays \p idx and \p dist, the set of nearest neighbor
+results are enumerated along dim0, in which the ith result is the ith nearest
+point to the query point. The result sets for the queries are placed in
+consecutive columns along dim1, in the same order as the queries are listed
+along dim1 of \p query. The output arrays therefore will have a shape of \p
+n_dist \f$\times\f$ the number of queries.
+
+For illustration, a simple example is given below for 1 query in 1-dimensional
+space. There are 6 points in \p train, and 3 of the nearest neighbors are
+queried for (\p n_dist is 3), so there are 3 elements in the result set for this
+single query, enumerated along dim0. In the results, \p idx describes that the 3
+points closest to 1.25 are, from nearest to farthest: point 0 in \p train (1.)
+with an SSD distance (as described in \p dist) of 0.0625 from the query, point 1
+(2.) with a distance of 0.5625, and point 2 (3.) with a distance of 3.0625.
+
+\snippet test/nearest_neighbour.cpp ex_nearest_1
+
+A slightly more complicated example is given below. There are 2 \p query points
+and 6 \p train points, and they are in 3-dimensional space (each point's
+coordinates are described along dim0, the list of points is enumerated along
+dim1). Note that in the output arrays \p idx and \p dist, there are 2 result
+sets now, one for each query. The result sets located on the first columns in
+the output arrays correspond to the first query (the first column in \p query),
+and that the result sets on the second columns of the output arrays correspond
+to the second query (second column in \p query). Thus, for example, the second
```

Comment:
by first and second `columns` do you mean all columns across slices ?

Reply:
No, because we don't have the notion of slices in this function. By first "columns" I mean the first column of `idx` and the first column of `dist`. In any case yeah this should be rephrased as it can be easily confusing.